sssd_test_framework/utils/adcli.py

```diff
@@ -0,0 +1,197 @@
+        def test_adcli_join(client: Client, provider: GenericADProvider):
+            cred = provider.host.adminpw
+            r = client.adcli.join(provider.host.domain, ["--domain-controller", provider.host.hostname], password=cred)
+            assert provider.host.domain in r.stdout, "adcli failed to join the client!"
+
+    .. note::
+
+       This utility will not revert any changes. It relies on AD host topology for clean up.
+       For methods requiring an authentication, --stdin-password(-W) is a default. Setting krb=True will enable
+       kerberos based authentication.
+    """
+
+    def info(self, domain: str, *, args: list[str] | None = None) -> ProcessResult:
+        """
+        Discover AD domain.
+
+        :param domain: domain.
+        :type domain: str
+        :param args: additional arguments, defaults to None
+        :type args: list[str] | None, optional
+        :return: Result of called command.
+        :rtype: ProcessResult
+        """
+        if args is None:
+            args = []
+
+        command = self.host.conn.exec(["adcli", "info", *args, "--verbose", f"{domain}"], raise_on_error=False)
+
+        return command
+
+    def testjoin(
+        self,
+        domain: str,
+        *,
+        args: list[str] | None = None,
+    ) -> ProcessResult:
+        """
+        Validate join.
+
+        :param domain: Domain.
+        :type domain: str
+        :param args: Additional arguments, defaults to None
+        :type args: list[str] | None, optional
+        :return: Result of called command.
+        :rtype: ProcessResult
+        """
+        if args is None:
+            args = []
+
+        return self.host.conn.exec(["adcli", "testjoin", "--verbose", domain, *args], raise_on_error=False)
+
+    def join(
+        self,
+        domain: str,
+        *,
+        args: list[str] | None = None,
+        password: str,
+        login_user: str,
+        krb: bool = False,
+    ) -> ProcessResult:
+        """
+        Create a computer account.
+
+        :param domain: Domain.
+        :type domain: str
+        :param args: Additional arguments, defaults to None
+        :type args: list[str] | None, optional
+        :param password: Password
+        :type password: str
+        :param login_user: Authenticating User
+        :type login_user: str
+        :param krb: Use Kerberos credentials, defaults to False
+        :type krb: bool, optional
+        :return: Result of called command.
+        :rtype: ProcessResult
+        """
+        if args is None:
+            args = []
+
+        if krb:
+            self.host.conn.exec(["kinit", f"{login_user}@{domain.upper()}"], input=password)
+            command = self.host.conn.exec(["adcli", "join", "--verbose", "-C", *args, domain], raise_on_error=False)
+        else:
+            command = self.host.conn.exec(
+                ["adcli", "join", "--verbose", f"--login-user={login_user}", *args, domain],
```

Comment:
From my testing, I think "--stdin-password" is needed here to take the password from `input=password`.